src/Lucene.Net/Index/ArrayHolder.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Buffers;
+using System.Threading;
+using Lucene.Net.Store;
+
+namespace Lucene.Net.Index
+{
+    public class ArrayHolder : IDisposable
+    {
+        private readonly int _size;
+        private readonly Directory _directory;
+        private readonly string _name;
+        private readonly long[] _longArray;
+        private readonly Term[] _termArray;
+        private readonly TermInfo[] _termInfoArray;
+        private int _usages;
+
+        public Span<long> LongArray => _longArray.AsSpan(0, _size);
+        public Span<TermInfo> InfoArray => _termInfoArray.AsSpan(0, _size);
+        public Span<Term> IndexTerms => _termArray.AsSpan(0, _size);
+
+        public ArrayHolder(int size, Directory directory, string name)
+        {
+            _size = size;
+            _directory = directory;
+            _name = name;
+            _longArray = ArrayPool<long>.Shared.Rent(size);
+            _termArray = ArrayPool<Term>.Shared.Rent(size);
+            _termInfoArray = ArrayPool<TermInfo>.Shared.Rent(size);
+        }
+
+        public void AddRef()
+        {
+            Interlocked.Increment(ref _usages);
+        }
+
+        public void ReleaseRef()
+        {
+            if (Interlocked.Decrement(ref _usages) == 0)
+                _directory.RemoveFromTermsIndexCache(_name);
+        }
+
+        public static ArrayHolder GenerateArrayHolder(Directory directory, string name, FieldInfos fieldInfos, int readBufferSize, int indexDivisor, IState state)
+        {
+            var indexEnum = new SegmentTermEnum(directory.OpenInput(name, readBufferSize, state), fieldInfos, true, state);
+
+            try
+            {
+                int indexSize = 1 + ((int)indexEnum.size - 1) / indexDivisor; // otherwise read index
+
+                var holder = new ArrayHolder(indexSize, directory, name);
+
+                for (int i = 0; indexEnum.Next(state); i++)
+                {
+                    holder.IndexTerms[i] = indexEnum.Term;
+                    holder.InfoArray[i] = indexEnum.TermInfo();
+                    holder.LongArray[i] = indexEnum.indexPointer;
+
+                    for (int j = 1; j < indexDivisor; j++)
+                        if (!indexEnum.Next(state))
+                            break;
+                }
+
+                return holder;
+
+            }
+            finally
+            {
+                indexEnum?.Close();
+            }
+        }
+
+        public void Dispose()
+        {
+            GC.SuppressFinalize(this);
+
+            if (_size > 64 * 1024)
+                return;
+
+            if (_longArray != null)
+                ArrayPool<long>.Shared.Return(_longArray);
+
+            if (_termArray != null)
+                ArrayPool<Term>.Shared.Return(_termArray, clearArray: true);
+
+            if (_termInfoArray != null)
+                ArrayPool<TermInfo>.Shared.Return(_termInfoArray);
+        }
+
+        ~ArrayHolder()
+        {
```

Comment:
Previously we had also:

```
#if DEBUG
                Console.WriteLine("Array holder is leaking...");
#endif
```

Maybe it could be valueable to keep it

Reply:
Before that, we were calling `Dispose` directly.
Now, since it's possible that an `ArrayHolder` is held by multiple `TermInfosReader`, I'm letting the finalizer do this job.

Reply:
In a working system (constant indexing) and/or querying this cache is kept alive forever (for each segment).
There are 3 cases in which we get rid of the cache:
- Index merge.
- When the system is idle, we cleanup the readers.
- When we are in low memory state, we cleanup the writer.

Since the cache is now shared between the writer and the readers I cannot `Dispose` it directly.